source/extensions/common/tap/tap_config_base.cc

```diff
@@ -13,23 +13,31 @@ namespace Extensions {
 namespace Common {
 namespace Tap {
 
-void Utility::addBufferToProtoBytes(envoy::data::tap::v2alpha::Body& output_bytes,
-                                    uint32_t max_buffered_bytes, const Buffer::Instance& data) {
+bool Utility::addBufferToProtoBytes(envoy::data::tap::v2alpha::Body& output_bytes,
+                                    uint32_t max_buffered_bytes, const Buffer::Instance& data,
+                                    uint32_t buffer_start_offset, uint32_t buffer_length_to_copy) {
   // TODO(mattklein123): Figure out if we can use the buffer API here directly in some way. This is
   // is not trivial if we want to avoid extra copies since we end up appending to the existing
   // protobuf string.
+  ASSERT(buffer_start_offset + buffer_length_to_copy <= data.length());
+  bool truncated = false;
```

Comment:
given the ASSERT above, this could be `truncated = output_bytes.as_bytes().size() + buffer_length_to_copy > max_buffered_bytes` here. Then it can calculate the exact number of bytes to copy before `trimSlices`, so the for loop can be simplified without involving `max_buffered_bytes` inside the loop.